3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala

```diff
@@ -274,12 +275,12 @@ class KinesisSink(provider: AWSCredentialsProvider,
           val failurePairs = unsentRecords zip results filter { _._2.getErrorMessage != null }
           info(s"Successfully wrote ${unsentRecords.size-failurePairs.size} out of ${unsentRecords.size} records")
           if (failurePairs.size > 0) {
-            failurePairs foreach { f => error(s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]") }
+            val (unsentRecords, failedResults) = failurePairs.unzip
+            logErrorsSummary(getErrorsSummary(failedResults))
             backoffTime = getNextBackoff(backoffTime)
             error(s"Retrying all failed records in $backoffTime milliseconds...")
 
             val err = s"Failed to send ${failurePairs.size} events"
-            unsentRecords = failurePairs.map(_._1)
```

Comment:
The dangers of mutable state: I forgot that I was meant to modify the unsentRecords var and instead created a new unsentRecords val shadowing it.